src/Microsoft.ML.TimeSeries/ExtensionsCatalog.cs

```diff
@@ -175,8 +177,8 @@ public static class TimeSeriesCatalog
         /// </format>
         /// </example>
         public static IDataView DetectEntireAnomalyBySrCnn(this AnomalyDetectionCatalog catalog, IDataView input, string outputColumnName, string inputColumnName,
-            double threshold = 0.3, int batchSize = 1024, double sensitivity = 99, SrCnnDetectMode detectMode = SrCnnDetectMode.AnomalyOnly)
-            => new SrCnnEntireAnomalyDetector(CatalogUtils.GetEnvironment(catalog), input, inputColumnName, outputColumnName, threshold, batchSize, sensitivity, detectMode);
+            double threshold = 0.3, int batchSize = 1024, double sensitivity = 99, SrCnnDetectMode detectMode = SrCnnDetectMode.AnomalyOnly, int period = 0, SrCnnDeseasonalityMode deseasonalityMode = SrCnnDeseasonalityMode.Stl)
+            => new SrCnnEntireAnomalyDetector(CatalogUtils.GetEnvironment(catalog), input, inputColumnName, outputColumnName, threshold, batchSize, sensitivity, detectMode, period, deseasonalityMode);
```

Comment:
So I believe that @harishsk has suggested you offline that instead of adding more parameters to this public API, you made another method that actually accepts an Options object containing the new parameters.

Examples on how we do it on ML.NET can be found here:
* In ImageClassification API methods: https://github.com/dotnet/machinelearning/blob/0fcf7ae132a4787cb6df414e4d8b49b35d509ff4/src/Microsoft.ML.Vision/VisionCatalog.cs#L87-L132
* In the LoadText API methods:
https://github.com/dotnet/machinelearning/blob/0fcf7ae132a4787cb6df414e4d8b49b35d509ff4/src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs#L135-L215

In both cases you'll find 2 methods, one which accepts several optional parameters, and another one which accepts an `Options `object. The reason to have 2 methods is because we originally created the version with the optional parameters first, but as time went by we realized we needed to introduce new parameters, and we created the Options method.

In your case, it will be necessary to maintain the existing `DetectEntireAnomalyBySrCnn `API with all the parameters it already had (without adding the new optional parameters you introduced: `period `and `deaseasonalityMode`). Then, you'll need to add another `DetectEntireAnomalyBySrCnn `method without optional parameters, but with an Options object as parameter... this new Options class will include your new parameters `period` and `deseasonalityMode`.

Reply:
Updated this function, however, there is a gap when I implement this option class. As I put it inside the SrCnnEntireAnomalyDetector class, this class can only be internal and thus can not be exposed in the ExtensionCatalog class. I have to put it outside SrCnnEntireAnomalyDetector. Is there any suggested methods to do it?

---
In reply to: [440466880](https://github.com/dotnet/machinelearning/pull/5202#discussion_r440466880) [](ancestors = 440466880)

Reply:
I see. Since `SrCnnEntireAnomalyDetector` is `internal`, and we can't make it `public` because it inherits from other `internal` classes, then I think your approach is ok and it's the best we can do :+1: (i.e. adding the new `public SrCnnEntireAnomalyDetectorOptions` class outside of the `internal SrCnnEntireAnomalyDetector` class)

---
In reply to: [440798983](https://github.com/dotnet/machinelearning/pull/5202#discussion_r440798983) [](ancestors = 440798983,440466880)

Reply:
Thanks for the clarification

---
In reply to: [441185662](https://github.com/dotnet/machinelearning/pull/5202#discussion_r441185662) [](ancestors = 441185662,440798983,440466880)